Design/spectrum_number_and_workspace_index_abstraction.md

```diff
@@ -0,0 +1,269 @@
+  - The remaining performance critical case is loading event data. Each event has an associated detector ID, and this needs to be translated to a workspace index.
+    This is currently done by creating a `vector` (instead of a `map`) for look-up.
+    We certainly have to support a similar mechanism.
+    This is a case where we will need to write some performance tests, to prevent regression.
+    These tests should be added well in advance so we can gather enough statistics prior to any changes and monitor the introduction of the changes.
+
+To deal with MPI in a way that is hidden from both the user and most developers, `Translator` transforms a user input into, e.g., a `SpectrumIndexSet`.
+This is best illustrated by the example from above:
+
+```python
+# User input: SpectrumNumbers='2,3,7'
+indices = ws.makeSpectrumIndexSet(getProperty('SpectrumNumbers'))
+```
+
+Consider 3 cases:
+
+```python
+# Without MPI (or single MPI rank)
+Spectrum number        1 2 3 7 8
+Workspace index        0 1 2 3 4
+Global workspace index 0 1 2 3 4
+# indices contains 1,2,3
+
+# 2 MPI ranks (interleaved distribution of spectra)
+Spectrum number        1 2 3 7 8
+MPI rank               0 1 0 1 0
+Workspace index        0 0 1 1 2
+Global workspace index 0 1 2 3 4
+# rank 0: indices contains 1
+# rank 1: indices contains 0,1
+
+# 3 MPI ranks (block-wise distribution of spectra)
+Spectrum number        1 2 3 7 8
+MPI rank               0 0 1 1 2
+Workspace index        0 1 0 1 0
+Global workspace index 0 1 2 3 4
+# rank 0: indices contains 1
+# rank 1: indices contains 0,1
+# rank 2: indices is empty
+```
+
+Algorithms then use:
+
+```python
+for index in indices:
+  ws.mutableY(index) += offset
+```
+
+There are a few more complex cases.
+For example, in `LoadEventNexus` a vector of events labeled by detector ID needs to be inserted into a workspace.
+This requires *filtering* a vector with data based on a spectrum number of detector ID.
+That is, we would not return a set of indices but a vector that is a subset of the input vector, i.e., the input vector with all data corresponding to other MPI ranks removed.
+
+#### Implementation and effects on existing code
+
+- Implement the `Translator` and store it in a workspace.
+  Most likely it makes sense to share this between workspaces, i.e., use `Kernel::cow_ptr`, since indices change rarely after initial creation.
+  As either part of the `Translator`, or as a separate entity within `MatrixWorkspace`, we will have a `Partitioning` object.
+  This object is responsible for determining how data of spectra is split among MPI ranks.
+  There can be several implementations of this `Partitioning` to fit the needs of various work flows.
+  `Translator` uses `Partitioning` to set up its internal maps that define which spectrum numbers are present on a specific MPI rank.
```

Comment:
Based on my thoughts above, I think it could be like this:

- `MatrixWorkspace` has `Translator` and `Partitioning` members.
- There are different `Partionings`, including a "default" one for no MPI/one rank.
- `MatrixWorkspace::setTranslator(....)` could, before storing, adapt the translator to the desired partitioning scheme using the partitioner, along the lines of `m_translator = m_partitioner.transformTranslator(userSuppliedTranslator);`.

Disadvantage: `MatrixWorkspace::getTranslator` does not necessarily return what you put into `MatrixWorkspace::setTranslator`.